### PR TITLE
Expand historical coverage pre-2019

### PR DESCRIPTION
**Pull request by grgmiller**
## Summary

This PR updates the data pipeline to allow for the creation of historical data from 2013-2018. Because EIA-930 data is not available for a complete year prior to 2019, the data outputs prior to that year will be limited to the following:
- Only monthly and annual-resolution data will be available (no hourly data) - we could potentially still include hourly data for the plants that report to CEMS.
- Only generated data will be available since interchange data from EIA-930 is not available.

## Where to look

Most of the updates are in `data_pipeline.py` with minor changes in other files to update allowed year ranges, and update certain functions to accept an argument to specify different behavior based on whether hourly data is available or not.

## Update details

Document in more depth the changes being made

## Screenshots

A couple screenshots of the changes/data if relevant.

## Testing / Validation

- [ ] Run the entire pipeline for 2018 without Errors, examine warning messages
- [ ] Run the entire pipeline for 2017 without Errors, examine warning messages
- [ ] Run the entire pipeline for 2016 without Errors, examine warning messages
- [ ] Run the entire pipeline for 2015 without Errors, examine warning messages
- [ ] Run the entire pipeline for 2014 without Errors, examine warning messages

After running the 2018 pipeline, I noticed the following warnings that are not tripped in the more recent data:
- Negative Nox emissions detected in CEMS
- Columns missing in the column_checks
- "Missing" data being dropped from cems with non-zero data
- SO2 emission factors appear to be missing for JF and certain boiler configs
- Some data is getting dropped or changed during the EIA-923 allocation process

## Linear ticket

Closes CAR-2968, CAR-1823, CAR-4206

## Concerns

Anything you'd like to point out that the reviewers should pay special attention to

## Next steps / Not addressed here

The availability of certain input data prior to 2013 may be different so that will be addressed in a future PR.

## Checklist

- [ ] Update the documentation to reflect changes made in this PR

### CODE REVIEW

**Comment by grgmiller** [comment]
Picking this PR back up on 11/22/23 after months of inactivity. At this point, just merged the most recent development branch in and did a test run of the pipeline with a single year (2018) of data to see if it is working. I just wanted to sync this branch with the most recent changes before we started all of our other updates so it doesn't get too far out of sync, but will probably pick work back up on this after the 2022 data update is complete.

**Comment by rouille** [comment]
This PR is now part of larger group of PRs that aim to update the data pipeline to allow for the creation of historical data from 2005-2018. All the PRs created to the expansion of the historical coverage will be merged into the `historical_coverage_feature` feature branch.

This PR allow to run the pipeline without error from 2008 to 2018. Not that the warnings have not been investigated yet and the outputs have not been validated. This PR simply fixes errors encountered when running 2008 - 2018.

Next steps:
* Remove `add_subplant_id` keyword in `oge.data_cleaning.clean_eia923`
* Add columns in the plant_attributes file such as lat/lon, nameplate capacity
* Handle 2005, 2006 and 2007 that years for which some of the pudl table retrieved `oge.data_cleaning.clean_eia923` come back empty
* Compress  output files to save disk space

File: src/oge/data_cleaning.py [hunk 1]
```diff
@@ -697,6 +699,34 @@ def clean_cems(year: int, small: bool, primary_fuel_table, subplant_emission_fac
     # load the CEMS data
     cems = load_data.load_cems_data(year)
 
+    for column in ["co2_mass_lb", "nox_mass_lb", "so2_mass_lb"]:
```
Comment (grgmiller) [inline src/oge/data_cleaning.py:702]: We should probably rename `bad_pollutant_data` to be more specific like `negative_emissions_data` if we're just testing for negative values. And maybe add an inline comment at the beginning like "treat negative emissions as bad data and replace with missing values.

We might want to make this a separate function in `validation` rather than just implement this inline? What do you think? Or we could just keep it here

File: src/oge/data_cleaning.py [hunk 2]
```diff
@@ -946,6 +976,64 @@ def assign_fuel_type_to_cems(cems, year, primary_fuel_table):
         filler_column="energy_source_code_epa",
     )
 
+    # as a final backstop, assign an energy_source_code based on the plant's primary
```
Comment (grgmiller) [inline src/oge/data_cleaning.py:979]: I'm wondering if this should be added as a separate function for ease of readibility.

File: src/oge/output_data.py [hunk 3]
```diff
@@ -93,7 +93,7 @@ def zip_results_for_s3(year):
     # move and rename the plant attributes files
     shutil.copy(
         f"{results_folder()}/{year}/plant_data/plant_static_attributes.csv",
-        f"{data_folder()}/s3_upload/plant_static_attributes_{year}.csv",
+        f"{data_folder()}/s3_upload/{year}_plant_static_attributes.csv",
```
Comment (grgmiller) [inline src/oge/output_data.py:96]: We need to keep the naming convention with the year at the end of the name - that is what the website is looking for.

File: src/oge/output_data.py [hunk 4]
```diff
@@ -636,3 +612,18 @@ def add_generated_emission_rate_columns(df):
                 path_prefix,
                 skip_outputs,
             )
+
+
+def add_generated_emission_rate_columns(df):
```
Comment (grgmiller) [inline src/oge/output_data.py:617]: Missing docstring. 

Also it looks like this removes the original behavior to fill missing emission values with zero. Why was that changed?
```
# where the rate is missing because of a divide by zero (i.e.
                        # net_generation_mwh is zero), replace the emission rate with
                        # zero. We want to keep all other NAs so that they get flagged
                        # by our validation checks since this indicates an unexpected
                        # issue
                        df.loc[df["net_generation_mwh"] == 0, col_name] = df.loc[
                            df["net_generation_mwh"] == 0, col_name
                        ].fillna(0)
```

File: src/oge/validation.py [hunk 5]
```diff
@@ -805,9 +799,6 @@ def test_emissions_adjustments(df, year):
                     [
                         "report_date",
                         "plant_id_eia",
-                        "generator_id",
-                        "prime_mover_code",
-                        "energy_source_code",
```
Comment (grgmiller) [inline src/oge/validation.py:810]: Instead of removing these entirely, could we just implement a flexible column handling like:
```
columns_to_show = ["report_date",
                        "plant_id_eia",
                        "generator_id",
                         "emissions_unit_id_epa",
                        "prime_mover_code",
                        "energy_source_code",]
...
bad_adjustment[[col for col in columns_to_show if col in bad_adjustment.columns] + [f"{pollutant}_mass_lb",f"{pollutant}_mass_lb_adjusted",]]
```

**Review by grgmiller — COMMENTED** [review]
See comments - some changes requested.

**Comment by grgmiller** [comment]
In addition to the next steps you listed above, it looks like we will need to figure out how to deal with the download_eia923() function since it will not work with some of the early data, and some of the functions that use those raw files may need alternative file handling in those earlier years.

**Comment by grgmiller** [comment]
Also, can you please test this to make sure that the results for say 2022 match the existing outputs? This change should theoretically not affect any of the results

File: src/oge/data_cleaning.py (hunk 1 above)
Comment (rouille) [reply src/oge/data_cleaning.py:702]: I don't have any preference. It is an edge case that happen only in cems so we could leave the logic here. We do the same thing in the `clean_eia923` function for `fuel_consumed_mmbtu` and `fuel_consumed_for_electricity_mmbtu`

File: src/oge/output_data.py (hunk 3 above)
Comment (rouille) [reply src/oge/output_data.py:96]: Good catch. It was part of the original commit and the rebase left it untouched

File: src/oge/output_data.py (hunk 4 above)
Comment (rouille) [reply src/oge/output_data.py:617]: It comes from the first commit: https://github.com/singularity-energy/open-grid-emissions/pull/295/commits/dd1b9203284a42b4ef70c9abfbcf3540ffceb97e

I did not make this edit.

File: src/oge/validation.py (hunk 5 above)
Comment (rouille) [reply src/oge/validation.py:810]: I will do that

File: src/oge/output_data.py (hunk 4 above)
Comment (rouille) [reply src/oge/output_data.py:617]: It looks like you factored out the `add_generated_emission_rate_columns` function from the ` write_power_sector_results` function and you refactored it by adding the `fillna(0)` right before the 2 `replace`. I believe it is equivalent than adding the extra assignment:
```
df.loc[df["net_generation_mwh"] == 0, col_name] = df.loc[
    df["net_generation_mwh"] == 0, col_name
].fillna(0)
```

Comment (grgmiller) [reply src/oge/output_data.py:617]: Ok maybe we leave the update then. I guess this would be consistent with having a missing emissions rate if there is no generation data to report, rather than a zero rate, which could be confused with zero-carbon

File: src/oge/validation.py (hunk 5 above)
Comment (rouille) [reply src/oge/validation.py:810]: Done

File: src/oge/data_cleaning.py (hunk 1 above)
Comment (rouille) [reply src/oge/data_cleaning.py:703]: Added comments and renamed variable but I did not create a separate function

File: src/oge/data_cleaning.py (hunk 2 above)
Comment (rouille) [reply src/oge/data_cleaning.py:980]: LMK if you want this to be done in this PR

File: src/oge/output_data.py (hunk 3 above)
Comment (rouille) [reply src/oge/output_data.py]: Changed

File: src/oge/output_data.py (hunk 4 above)
Comment (rouille) [reply src/oge/output_data.py]: Changed to what we have in `development`/`main`. As a result, the warnings that were raised for some BA, such as:
```
2024-05-14 12:23:37 [INFO] oge.oge.validation:222 Checking that fuel and emissions values are positive...  
2024-05-14 12:23:37 [INFO] oge.oge.validation:296 OK
2024-05-14 12:23:37 [INFO] oge.oge.validation:302 Checking that no values are missing...  
2024-05-14 12:23:37 [WARNING] oge.oge.validation:308 There are 4 records where generated_co2_rate_lb_per_mwh_for_electricity is missing.
2024-05-14 12:23:37 [WARNING] oge.oge.validation:308 There are 4 records where generated_co2e_rate_lb_per_mwh_for_electricity is missing.
2024-05-14 12:23:37 [WARNING] oge.oge.validation:308 There are 4 records where generated_nox_rate_lb_per_mwh_for_electricity is missing.
2024-05-14 12:23:37 [WARNING] oge.oge.validation:308 There are 4 records where generated_so2_rate_lb_per_mwh_for_electricity is missing.
2024-05-14 12:23:37 [WARNING] oge.oge.validation:308 There are 4 records where generated_co2_rate_lb_per_mwh_for_electricity_adjusted is missing.
2024-05-14 12:23:37 [WARNING] oge.oge.validation:308 There are 4 records where generated_co2e_rate_lb_per_mwh_for_electricity_adjusted is missing.
2024-05-14 12:23:37 [WARNING] oge.oge.validation:308 There are 4 records where generated_nox_rate_lb_per_mwh_for_electricity_adjusted is missing.
2024-05-14 12:23:37 [WARNING] oge.oge.validation:308 There are 4 records where generated_so2_rate_lb_per_mwh_for_electricity_adjusted is missing.
2024-05-14 12:23:37 [WARNING] oge.oge.validation:318 The above missing values are errors and must be fixed
```
now disappeared

**Comment by rouille** [comment]
> In addition to the next steps you listed above, it looks like we will need to figure out how to deal with the download_eia923() function since it will not work with some of the early data, and some of the functions that use those raw files may need alternative file handling in those earlier years.

Indeed:
```
open-grid-emissions[~/Singularity/open-grid-emissions/src/oge] (historical_coverage) brdo$ python data_pipeline.py --year 2005
2024-05-14 15:26:48 [INFO] oge.data_pipeline:71 

Running with the following options:
  * year = 2005
  * shape_individual_plants = True
  * small = False
  * flat = False
  * skip_outputs = False

2024-05-14 15:26:48 [INFO] oge.data_pipeline:121 Running data pipeline for year 2005
2024-05-14 15:26:48 [WARNING] oge.oge.validation:32 
        ################################################################################
        The data pipeline has only been validated to work for years 2019-2022.
        Running the pipeline for 2005 may cause it to fail or may lead to poor-quality
        or anomalous results. To check on the progress of validating additional years of
        data, see: https://github.com/singularity-energy/open-grid-emissions/issues/117
        ################################################################################
        
2024-05-14 15:26:48 [INFO] oge.data_pipeline:126 1. Downloading data
2024-05-14 15:26:48 [INFO] oge.oge.download_data:126 Using nightly build version of PUDL sqlite database downloaded 2024-04-03
2024-05-14 15:26:48 [INFO] oge.oge.download_data:147 Using nightly build version of PUDL epacems parquet file downloaded 2024-04-03
2024-05-14 15:26:48 [INFO] oge.oge.download_data:45 egrid2018_data.xlsx already downloaded, skipping.
2024-05-14 15:26:48 [INFO] oge.oge.download_data:45 egrid2019_data.xlsx already downloaded, skipping.
2024-05-14 15:26:48 [INFO] oge.oge.download_data:45 egrid2020_data.xlsx already downloaded, skipping.
2024-05-14 15:26:48 [INFO] oge.oge.download_data:45 egrid2021_data.xlsx already downloaded, skipping.
2024-05-14 15:26:48 [INFO] oge.oge.download_data:45 egrid2022_data.xlsx already downloaded, skipping.
2024-05-14 15:26:48 [INFO] oge.oge.download_data:45 epa_eia_crosswalk.csv already downloaded, skipping.
2024-05-14 15:26:48 [INFO] oge.oge.download_data:45 eia8602005 already downloaded, skipping.
2024-05-14 15:26:48 [INFO] oge.oge.download_data:45 eia8602022 already downloaded, skipping.
Traceback (most recent call last):
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/data_pipeline.py", line 656, in <module>
    main(sys.argv[1:])
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/data_pipeline.py", line 147, in main
    download_data.download_raw_eia923(year)
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/download_data.py", line 302, in download_raw_eia923
    raise NotImplementedError(f"EIA-923 data is unavailable for '{year}'.")
NotImplementedError: EIA-923 data is unavailable for '2005'.
```

File: src/oge/data_cleaning.py (hunk 2 above)
Comment (grgmiller) [reply src/oge/data_cleaning.py:980]: Sure, let's go ahead and do this so it doesn't get lost.

Comment (rouille) [reply src/oge/data_cleaning.py]: Done. LMK what you think. Happy to change function name, etc.

File: src/oge/data_cleaning.py [hunk 6]
```diff
@@ -946,12 +977,81 @@ def assign_fuel_type_to_cems(cems, year, primary_fuel_table):
...
+        filler_column="energy_source_code_plant",
+    )
+
     # update
     cems = update_energy_source_codes(cems, year)
 
     return cems
 
 
+def assign_fuel_type_via_generator_with_greatest_capacity() -> pd.DataFrame:
```
Comment (grgmiller) [inline src/oge/data_cleaning.py:1001]: I might suggest renaming this something like `load_backstop_energy_source_codes_for_plant()`. Otherwise looks good

**Comment by grgmiller** [comment]
I added one comment with a suggested name change, otherwise this looks good to merge once we confirm that this is not modifying the 2022 outputs.

File: src/oge/data_cleaning.py (hunk 6 above)
Comment (rouille) [reply src/oge/data_cleaning.py]: Done

**Comment by grgmiller** [comment]
@rouille Looks good to me - ready to merge!

**Review by rouille — APPROVED** [review]
